From: Ben Hutchings <ben@decadent.org.uk>
Date: Wed, 5 Jul 2017 13:32:43 +0100
Subject: mmap: Remember the MAP_FIXED flag as VM_FIXED

Backport to 3.2: there are no spare bits, but we can share with VM_SAO
as that's only used on powerpc and VM_FIXED will only be needed on
x86.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
---
 include/linux/mm.h   | 1 +
 include/linux/mman.h | 3 ++-
 2 files changed, 3 insertions(+), 1 deletion(-)

--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -115,7 +115,13 @@ extern unsigned int kobjsize(const void
 
 #define VM_CAN_NONLINEAR 0x08000000	/* Has ->fault & does nonlinear pages */
 #define VM_MIXEDMAP	0x10000000	/* Can contain "struct page" and pure PFN pages */
+#ifdef CONFIG_PPC
 #define VM_SAO		0x20000000	/* Strong Access Ordering (powerpc) */
+#define VM_FIXED	0x00000000
+#else
+#define VM_SAO		0x00000000
+#define VM_FIXED	0x20000000	/* Allocated at fixed address */
+#endif
 #define VM_PFN_AT_MMAP	0x40000000	/* PFNMAP vma that is fully mapped at mmap time */
 #define VM_MERGEABLE	0x80000000	/* KSM may merge identical pages */
 
--- a/include/linux/mman.h
+++ b/include/linux/mman.h
@@ -87,7 +87,9 @@ calc_vm_flag_bits(unsigned long flags)
 	return _calc_vm_trans(flags, MAP_GROWSDOWN,  VM_GROWSDOWN ) |
 	       _calc_vm_trans(flags, MAP_DENYWRITE,  VM_DENYWRITE ) |
 	       _calc_vm_trans(flags, MAP_EXECUTABLE, VM_EXECUTABLE) |
-	       _calc_vm_trans(flags, MAP_LOCKED,     VM_LOCKED    );
+	       _calc_vm_trans(flags, MAP_LOCKED,     VM_LOCKED    ) |
+	       (VM_FIXED ?
+		_calc_vm_trans(flags, MAP_FIXED,     VM_FIXED     ) : 0);
 }
 #endif /* __KERNEL__ */
 #endif /* _LINUX_MMAN_H */
